samza-core/src/main/java/org/apache/samza/drain/DrainMonitor.java

```diff
@@ -0,0 +1,261 @@
+        LOG.info("Started DrainMonitor.");
+        break;
+      case RUNNING:
+      case DRAINING:
+      case STOPPED:
+        LOG.warn("Cannot call start() on the DrainMonitor when it is in {} state.", state);
+        break;
+    }
+  }
+
+  /**
+   * Stops the DrainMonitor.
+   *
+   * @param shouldClearDrainMessages parameter to indicate if the DrainMonitor should perform cleanup of old
+   *                                 drain messages.
+   * */
+  public void stop(boolean shouldClearDrainMessages) {
+    switch (state) {
+      case RUNNING:
+        schedulerService.shutdownNow();
+        state = State.STOPPED;
+        LOG.info("Stopped DrainMonitor.");
+        break;
+      case INIT:
+      case STOPPED:
+      case DRAINING:
+        // no-op
+        LOG.info("Cannot call stop() on the DrainMonitor when it is in {} state.", state);
+        break;
+    }
+    if (shouldClearDrainMessages) {
+      deleteAllDrainNotifications();
+    }
+  }
+
+  private void drain() {
+    switch (state) {
+      case RUNNING:
+        schedulerService.shutdownNow();
+        state = State.DRAINING;
+        LOG.info("Stopped DrainMonitor.");
+        break;
+      case INIT:
+      case STOPPED:
+      case DRAINING:
+        LOG.info("Cannot call drain() on the DrainMonitor when it is in {} state.", state);
+        break;
+    }
+  }
+
+  /**
+   * Register a callback to be executed when DrainNotification is encountered.
+   *
+   * @param callback the callback to register.
+   * @return Returns {@code true} if registration was successful and {@code false} if not.
+   * Registration can fail it the DrainMonitor is stopped or a callback is already registered.
+   * */
+  public boolean registerDrainCallback(DrainCallback callback) {
+    Preconditions.checkNotNull(callback);
+
+    switch (state) {
+      case RUNNING:
+      case DRAINING:
+      case STOPPED:
+        LOG.warn("Cannot register callback when it is in {} state. Please register callback before calling start "
+            + "on DrainMonitor.", state);
+        return false;
+      case INIT:
+        if (this.callback != null) {
+          LOG.warn("Cannot register callback as a callback is already registered.");
+          return false;
+        }
+        this.callback = callback;
+        return true;
+      default:
+        return false;
+    }
+  }
+
+  /**
+   * Callback for any action to executed by DrainMonitor implementations once Drain is encountered.
+   * Registered using {@link #registerDrainCallback(DrainCallback)}.
+   * */
+  public interface DrainCallback {
+    void onDrain();
+  }
+
+  /**
+   * Check if the DrainMonitor is in Draining state.
+   * */
+  public boolean isDraining() {
+    return state == State.DRAINING;
+  }
+
+  /**
+   * One time check check to see if there are any DrainNotification messages available in the
+   * metadata store for the current deployment.
+   * */
+  public boolean shouldDrain() {
```

Comment:
private?

Reply:
Thanks, made this package-private. Previously it was required to be public as it was used by the SamzaContainer directly